In [1]:
# Install and import necessary libraries
!pip install spotipy
!pip install pylast
import requests
import pandas as pd
import numpy as np

In [2]:
# Load DataFrame from pickle file
songs_final_df = pd.read_pickle('songs_final_df.pkl')

In [3]:
# Show the entire DataFrame
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
songs_final_df

track_name  \
0                                      Blinding Lights   
1                                         Shape of You   
2                                    Someone You Loved   
3                                         Dance Monkey   
4        Sunflower - Spider-Man: Into the Spider-Verse   
5                                            One Dance   
6                            STAY (with Justin Bieber)   
7                                             Believer   
8                                               Closer   
9                                              Starboy   
10                                             Perfect   
11                                          Heat Waves   
12                                           As It Was   
13                                           Se��o   
14                                Say You Won't Let Go   
15                                lovely - Bonus Track   
16                                    Watermelon Sugar   
17                                     Don't Start Now   
18                                        Lucid Dreams   
19                                     Sweater Weather   
20                                   Thinking Out Loud   
21                                          Photograph   
22                            Something Just Like This   
23                 Bohemian Rhapsody - Remastered 2011   
24                                             Shallow   
25                                   Take Me To Church   
26                                             Circles   
27                                       Love Yourself   
28                                           All of Me   
29                                      Counting Stars   
30                                             Riptide   
31                             Wake Me Up - Radio Edit   
32                    Can't Hold Us (feat. Ray Dalton)   
33                                           The Hills   
34                                             HUMBLE.   
35                            One Kiss (with Dua Lipa)   
36                                            good 4 u   
37                                     drivers license   
38                                              Demons   
39                  Lose Yourself - Soundtrack Version   
40                   INDUSTRY BABY (feat. Jack Harlow)   
41                                        Another Love   
42                                      Mr. Brightside   
43                           Levitating (feat. DaBaby)   
44                                      No Role Modelz   
45                                    Do I Wanna Know?   
46                                              Dakiti   
47                                            Memories   
48                                              Yellow   
49                      MONTERO (Call Me By Your Name)   
50                     There's Nothing Holdin' Me Back   
51                              Mood (feat. Iann Dior)   
52                                    'Till I Collapse   
53                                            Dynamite   
54           Smells Like Teen Spirit - Remastered 2021   
55                                          Without Me   
56                                 When I Was Your Man   
57                                         Die For You   
58                                Just The Way You Are   
59                                          In The End   
60                                       The Scientist   
61                                       Before You Go   
62                                I'm Not The Only One   
63                             Cold Heart - PNAU Remix   
64             Every Breath You Take - Remastered 2003   
65                                        Viva La Vida   
66                                     Save Your Tears   
67                            Kiss Me More (feat. SZA)   
68                                        

# Conclusions based on variables

## Column released_year + released_month

### Range of release year and month

In [4]:
# Find the earliest release date
min_date = songs_final_df[['released_year', 'released_month']].astype(str).agg('-'.join, axis=1).min()

# Find the latest release date
max_date = songs_final_df[['released_year', 'released_month']].astype(str).agg('-'.join, axis=1).max()

print("The date range goes from", min_date, "to", max_date)

The date range goes from 1942-1 to 2023-5


### Percentage of songs released in 2023

In [5]:
# Check how many songs were released in 2023 
songs_2023 = songs_final_df[songs_final_df['released_year'] == 2023].shape[0]

# Calculate total songs 
total_songs = songs_final_df.shape[0]

# Calculate percentage of songs released in 2023 
percentage_2023 = (songs_2023 / total_songs) * 100

print("The percentage of songs released in 2023 is:", percentage_2023, "%")


The percentage of songs released in 2023 is: 3.5999999999999996 %


### Most common year

In [6]:
# Count the frequency of each release year
years_frequency = songs_final_df['released_year'].value_counts()

# Get the most common year (highest frequency)
most_common_year = years_frequency.idxmax()

# Get the frequency of the most common year
most_common_frequency = years_frequency.max()

print("The most common year where songs were released is:", most_common_year)
print("Number of released songs in the most common year:", most_common_frequency)

The most common year where songs were released is: 2022
Number of released songs in the most common year: 162


### Percentage of songs released in 2022

In [7]:
# Check how many songs were released in 2022 
songs_2022 = songs_final_df[songs_final_df['released_year'] == 2022].shape[0]

# Calculate total songs 
total_2022_songs = songs_final_df.shape[0]

# Calculate percentage of songs released in 2022 
percentage_2022 = (songs_2022 / total_songs) * 100

print("The percentage of songs released in 2022 is:", percentage_2022, "%")

The percentage of songs released in 2022 is: 32.4 %


### Percentage of songs from the XX century

In [8]:
# Count number of songs from XX century
songs_xx = songs_final_df[(songs_final_df['released_year'] >= 1901) & (songs_final_df['released_year'] <= 2000)].shape[0]

# Calculate total of songs
total_songs = songs_final_df.shape[0]

# Calculate percentage of songs from XX Century
percentage_xx = (songs_xx / total_songs) * 100

print("The percentage of songs from the 20th century is:", percentage_xx, "%")

The percentage of songs from the 20th century is: 8.4 %


### Percentage of songs from 2000-2010

In [9]:
# Count number of songs from 2000 to 2010
songs_xx = songs_final_df[(songs_final_df['released_year'] >= 2000) & (songs_final_df['released_year'] <= 2010)].shape[0]

# Calculate total of songs
total_songs = songs_final_df.shape[0]

# Calculate percentage of songs from 2000 to 2010
percentage_xx = (songs_xx / total_songs) * 100

print("The percentage of songs from 2000 to 2010 is:", percentage_xx, "%")

The percentage of songs from 2000 to 2010 is: 5.0 %


### Percentage of songs from 2010-2023

In [10]:
# Count number of songs from 2010 to 2023
songs_xx = songs_final_df[(songs_final_df['released_year'] >= 2010) & (songs_final_df['released_year'] <= 2023)].shape[0]

# Calculate total of songs
total_songs = songs_final_df.shape[0]

# Calculate percentage of songs from 2010 to 2023
percentage_xx = (songs_xx / total_songs) * 100

print("The percentage of songs from 2010 to 2023 is:", percentage_xx, "%")

The percentage of songs from 2010 to 2023 is: 88.6 %


### Most common month

In [11]:
# Count the frequency of each release month
months_frequency = songs_final_df['released_month'].value_counts()

# Get the most common year (highest frequency)
most_common_month = months_frequency.idxmax()

# Get the frequency of the most common year
most_common_month_frequency = months_frequency.max()

print("The most common month where songs were released is:", most_common_month)
print("Number of released songs in the most common month:", most_common_month_frequency)

The most common month where songs were released is: 1
Number of released songs in the most common month: 88


## Streaming platforms

### Spotify playlists

In [12]:
# Find the index of the song that is in more playlists
index_max_playlists = songs_final_df['in_spotify_playlists'].idxmax()

# Find the name of the song that is in more playlists
song_max_playlists = songs_final_df.loc[index_max_playlists, 'track_name']

# Get the number of plyalists that song is in
max_playlists = songs_final_df.loc[index_max_playlists, 'in_spotify_playlists']

print("The song that is in more Spotify playlists is:", song_max_playlists)
print("Number of playlists it's in:", max_playlists)

The song that is in more Spotify playlists is: Get Lucky - Radio Edit
Number of playlists it's in: 52898


In [13]:
# Find the index of the song that is in less playlists
index_min_playlists = songs_final_df['in_spotify_playlists'].idxmin()

# Find the name of the song that is in less playlists
song_min_playlists = songs_final_df.loc[index_min_playlists, 'track_name']

# Get the number of playlists that song is in
min_playlists = songs_final_df.loc[index_min_playlists, 'in_spotify_playlists']

print("The song that is in less Spotify playlists is:", song_min_playlists)
print("Number of playlists it's in:", min_playlists)

The song that is in less Spotify playlists is: Dream On
Number of playlists it's in: 168


### Spotify Charts

In [14]:
# Find the index of the song with highest position in the chart
index_min_spoti_chart = songs_final_df['in_spotify_charts'].idxmin()

# Find the name of the song in the higer position
song_max_spoti_chart = songs_final_df.loc[index_min_spoti_chart, 'track_name']

# Get the position of that song in the chart
max_spoti_chart = songs_final_df.loc[index_min_spoti_chart, 'in_spotify_charts']

print("The song with the highest position in Spotify charts is:", song_max_spoti_chart)
print("Maximum position:", max_spoti_chart)

The song with the highest position in Spotify charts is: Dance Monkey
Maximum position: 0


### Apple Charts

In [15]:
# Find the index of the song with highest position in the chart
index_min_apple_chart = songs_final_df['in_apple_charts'].idxmin()

# Find the name of the song in the higer position
song_max_apple_chart = songs_final_df.loc[index_min_apple_chart, 'track_name']

# Get the position of that song in the chart
max_apple_chart = songs_final_df.loc[index_min_apple_chart, 'in_apple_charts']

print("The song with the highest position in Apple charts is:", song_max_apple_chart)
print("Maximum position:", max_apple_chart)

The song with the highest position in Apple charts is: Shape of You
Maximum position: 0


### Deezer Charts

In [16]:
# Find the index of the song with highest position in the chart
index_min_deezer_chart = songs_final_df['in_deezer_charts'].idxmin()

# Find the name of the song in the higer position
song_max_deezer_chart = songs_final_df.loc[index_min_deezer_chart, 'track_name']

# Get the position of that song in the chart
max_deezer_chart = songs_final_df.loc[index_min_deezer_chart, 'in_deezer_charts']

print("The song with the highest position in Deezer charts is:", song_max_deezer_chart)
print("Maximum position:", max_deezer_chart)

The song with the highest position in Deezer charts is: Someone You Loved
Maximum position: 0


### Shazam Charts

In [17]:
# Convert in_shazam_charts column to numeric type to handle NaN values
songs_final_df['in_shazam_charts'] = pd.to_numeric(songs_final_df['in_shazam_charts'], errors='coerce')

# Encontrar el índice de la canción con la posición más baja en el chart de Shazam
index_min_shazam_chart = songs_final_df['in_shazam_charts'].idxmin()

# Encontrar el nombre de la canción en la posición más alta
song_max_shazam_chart = songs_final_df.loc[index_min_shazam_chart, 'track_name']

# Obtener la posición más alta en el chart de Shazam
max_shazam_chart = songs_final_df.loc[index_min_shazam_chart, 'in_shazam_charts']

print("The song with the highest position in Shazam charts is:", song_max_shazam_chart)
print("Maximum position:", max_shazam_chart)

The song with the highest position in Shazam charts is: Shape of You
Maximum position: 0.0


## Column artist_count

In [18]:
# Calculate percentages of every value in the artist_count column
percentages_artist_count = songs_final_df['artist_count'].value_counts(normalize=True) * 100
percentages_artist_count

artist_count
1    67.0
2    23.8
3     7.8
4     1.0
7     0.2
5     0.2
Name: proportion, dtype: float64

## Streams

### Song with more streams

In [19]:
# Convert the streams column into numeric type
songs_final_df['streams'] = pd.to_numeric(songs_final_df['streams'], errors='coerce')

# Find the index of the song with more streams
index_max_streams = songs_final_df['streams'].idxmax()

# Get the name of the song with more streams
song_max_streams = songs_final_df.loc[index_max_streams, 'track_name']

# Get the number of streams for that song
max_streams = songs_final_df.loc[index_max_streams, 'streams']

print("The song with most streams is:", song_max_streams)
print("Number of streams:", max_streams)

The song with most streams is: Blinding Lights
Number of streams: 3703895074


### Song with less streams

In [20]:
# Find the index of the song with less streams
index_min_streams = songs_final_df['streams'].idxmin()

# Get the name of the song with less streams
song_min_streams = songs_final_df.loc[index_min_streams, 'track_name']

# Get the number of streams for that song
min_streams = songs_final_df.loc[index_min_streams, 'streams']

print("The song with less streams is:", song_min_streams)
print("Number of streams:", min_streams)

The song with less streams is: Cooped Up (with Roddy Ricch)
Number of streams: 271666301


In [21]:
songs_final_df['streams'].describe()

count    5.000000e+02
mean     8.530195e+08
std      6.048670e+08
min      2.716663e+08
25%      3.938303e+08
50%      6.322926e+08
75%      1.160155e+09
max      3.703895e+09
Name: streams, dtype: float64

### Top 10 songs with more streams

In [22]:
# Get the top 10 songs regarding streams
top_10_songs = songs_final_df.head(10)
top_10_songs

track_name                track_id  \
0                                Blinding Lights  0VjIjW4GlUZAMYd2vXMi3b   
1                                   Shape of You  7qiZfU4dY1lWllzX7mPBI3   
2                              Someone You Loved  7qEHsqek33rTcFNT9PFqLf   
3                                   Dance Monkey  2XU0oxnq2qxCpomAAuJY8K   
4  Sunflower - Spider-Man: Into the Spider-Verse  0RiRZpuVRbi7oqRdSMwhQY   
5                                      One Dance  1zi7xx7UVEFkmKfv06H8x0   
6                      STAY (with Justin Bieber)  5HCyWlXZPP0y6Gqq8TgA20   
7                                       Believer  0pqnGHJpmpxLKifKRmU6WP   
8                                         Closer  7BKLCZ1jbUBVqRi2FVlTVw   
9                                        Starboy  7MXVkk9YMctZqd1Srtv4MB   

      artist_1_name   artist_2_name artist_3_name artist_4_name artist_5_name  \
0        The Weeknd            None          None          None          None   
1        Ed Sheeran            None          None          None          None   
2     Lewis Capaldi            None          None          None          None   
3       Tones and I            None          None          None          None   
4       Post Malone        Swae Lee          None          None          None   
5             Drake          WizKid          Kyla          None          None   
6     Justin Bieber   The Kid Laroi          None          None          None   
7   Imagine Dragons            None          None          None          None   
8  The Chainsmokers          Halsey          None          None          None   
9        The Weeknd       Daft Punk          None          None          None   

   artist_count                                         artist_ids  \
0             1                             1Xyo4u8uXC1ZmMpatF05PJ   
1             1                             6eUKZXaKkcviH0Ku9w2n3V   
2             1                             4GNC7GD6oZMSxPGyXy4MNB   
3             1                             2NjfBq1NflQcKSeiDooVjY   
4             2     246dkjvS1zLTtiykXe5h60, 1zNqQNIdeOUZHb8zbZRFMX   
5             3  3TVXtAsR1Inumwj472S9r4, 3tVQdUvClmAT7URs9V3rsp...   
6             2     1uNFoZAHBGtllmzznpCI3s, 284HyOro5vrr8DXPrdPSQL   
7             1                             53XhwfbYqKCa1cC15pYq2q   
8             2     69GGBxA162lTqCwzJG5jLp, 26VFTg2z8YR0cCuwLzESi2   
9             2     1Xyo4u8uXC1ZmMpatF05PJ, 7CTejuKlwb0KuJQP80OAOZ   

   released_year  released_month  released_day  \
0           2019              11            29   
1           2017               1             6   
2           2018              11             8   
3           2019               5            10   
4           2018              10             9   
5           2016               4             4   
6           2021               7             9   
7           2017               1            31   
8           2016               5            31   
9           2016               9            21   

                                          genre  \
0  canadian contemporary r&b, canadian pop, pop   
1            pop, singer-songwriter pop, uk pop   
2                                   pop, uk pop   
3                                australian pop   
4                                          None   
5                                          None   
6                                          None   
7                        modern rock, pop, rock   
8                                          None   
9                                          None   

                                         album  in_spotify_playlists  \
0                                  After Hours                 43899   
1                                   ÷ (Deluxe)                 32181   
2      Divinely Uninspired To A Hellish Extent                 17836   
3  Dance Monkey (Stripped Back) / Dance Monkey                 24529   
4                         Hollywood's Bleeding                 24

### Top 10 Songs genre

In [23]:
top_10_songs['genre'].value_counts()

genre
None                                            5
canadian contemporary r&b, canadian pop, pop    1
pop, singer-songwriter pop, uk pop              1
pop, uk pop                                     1
australian pop                                  1
modern rock, pop, rock                          1
Name: count, dtype: int64

### Top 10 Songs artist_count

In [24]:
top_10_songs['artist_count'].value_counts()

artist_count
1    5
2    4
3    1
Name: count, dtype: int64

### Top 10 Songs duration

In [25]:
top_10_songs['duration'].describe()

count     10.00000
mean     197.84600
std       33.87831
min      141.80500
25%      176.02950
50%      202.19300
75%      225.19925
max      244.96000
Name: duration, dtype: float64

### Top 10 Songs attributes

In [26]:
top_10_songs['bpm'].describe()

count     10.00000
mean     124.50000
std       36.84879
min       90.00000
25%       96.50000
50%      107.00000
75%      158.75000
max      186.00000
Name: bpm, dtype: float64

In [27]:
top_10_songs['danceability_%'].describe()

count    10.000000
mean     69.700000
std      12.454807
min      50.000000
25%      61.250000
50%      75.500000
75%      77.000000
max      83.000000
Name: danceability_%, dtype: float64

In [28]:
top_10_songs['valence_%'].describe()

count    10.000000
mean     59.200000
std      20.660214
min      36.000000
25%      45.750000
50%      51.500000
75%      71.500000
max      93.000000
Name: valence_%, dtype: float64

In [29]:
top_10_songs['energy_%'].describe()

count    10.000000
mean     62.300000
std      12.858633
min      41.000000
25%      53.750000
50%      61.000000
75%      73.250000
max      80.000000
Name: energy_%, dtype: float64

In [30]:
top_10_songs['acousticness_%'].describe()

count    10.000000
mean     32.200000
std      30.308781
min       0.000000
25%       4.000000
50%      28.500000
75%      57.000000
max      75.000000
Name: acousticness_%, dtype: float64

In [31]:
top_10_songs['instrumentalness_%'].describe()

count    10.0
mean      0.0
std       0.0
min       0.0
25%       0.0
50%       0.0
75%       0.0
max       0.0
Name: instrumentalness_%, dtype: float64

In [32]:
top_10_songs['liveness_%'].describe()

count    10.000000
mean     14.700000
std       8.882567
min       7.000000
25%       9.250000
50%      11.000000
75%      16.750000
max      36.000000
Name: liveness_%, dtype: float64

In [33]:
top_10_songs['speechiness_%'].describe()

count    10.000000
mean      8.500000
std       7.367345
min       3.000000
25%       5.000000
50%       6.000000
75%       9.500000
max      28.000000
Name: speechiness_%, dtype: float64

### Common data with Top 10 songs

In [34]:
# Examine other columns in the DataFrame to identify common data
for column in top_10_songs.columns:
    # Si la columna no es 'streams' (la columna que contiene el número de streams)
    if column != 'streams':
        # Obtener los valores más comunes en la columna para las canciones en el top 10 de streams
        common_data = top_10_songs[column].value_counts()
        # Imprimir solo los valores que tienen más de una ocurrencia (es decir, los valores comunes)
        print(f"Common data in the column: '{column}':")
        print(common_data[common_data > 1])
        print() 

Common data in the column: 'track_name':
Series([], Name: count, dtype: int64)

Common data in the column: 'track_id':
Series([], Name: count, dtype: int64)

Common data in the column: 'artist_1_name':
artist_1_name
The Weeknd    2
Name: count, dtype: int64

Common data in the column: 'artist_2_name':
Series([], Name: count, dtype: int64)

Common data in the column: 'artist_3_name':
Series([], Name: count, dtype: int64)

Common data in the column: 'artist_4_name':
Series([], Name: count, dtype: int64)

Common data in the column: 'artist_5_name':
Series([], Name: count, dtype: int64)

Common data in the column: 'artist_count':
artist_count
1    5
2    4
Name: count, dtype: int64

Common data in the column: 'artist_ids':
Series([], Name: count, dtype: int64)

Common data in the column: 'released_year':
released_year
2016    3
2019    2
2017    2
2018    2
Name: count, dtype: int64

Common data in the column: 'released_month':
released_month
11    2
1     2
5     2
Name: count, dtype: int

## Songs attributes

### bpm

In [35]:
songs_final_df['bpm'].describe()

count    500.000000
mean     122.588000
std       29.354209
min       65.000000
25%       98.000000
50%      120.000000
75%      141.000000
max      206.000000
Name: bpm, dtype: float64

#### The average bpm is: 122.58. Songs in the DataFrame tend to have a higher tempo.

### key

In [36]:
songs_final_df['key'].value_counts()

key
C#    67
G#    49
G     45
F     44
D     42
B     42
F#    37
A     35
A#    32
E     32
D#    17
Name: count, dtype: int64

#### The predominant key is C#

### mode

In [37]:
songs_final_df['mode'].value_counts()

mode
Major    297
Minor    203
Name: count, dtype: int64

#### Most of the songs are in Mayor mode

### danceability_%

In [38]:
songs_final_df['danceability_%'].describe()

count    500.000000
mean      65.866000
std       14.937144
min       23.000000
25%       56.000000
50%       68.000000
75%       77.000000
max       95.000000
Name: danceability_%, dtype: float64

#### The average danceability is 65.86. Songs in the DataFrame tend to be more danceable.

### valence_%

In [39]:
songs_final_df['valence_%'].describe()

count    500.000000
mean      50.264000
std       23.193322
min        4.000000
25%       31.750000
50%       50.000000
75%       68.000000
max       97.000000
Name: valence_%, dtype: float64

#### The average valence is 50.26. Half of the songs in the DataFrame transmit positive vibes

### energy_%

In [40]:
songs_final_df['energy_%'].describe()

count    500.000000
mean      63.500000
std       16.242217
min       15.000000
25%       53.000000
50%       65.000000
75%       76.000000
max       97.000000
Name: energy_%, dtype: float64

#### The average energy is 63.5. Songs in the Dataframe tend to be energetic.

### acousticness_%

In [41]:
songs_final_df['acousticness_%'].describe()

count    500.000000
mean      26.490000
std       26.254075
min        0.000000
25%        5.000000
50%       16.000000
75%       42.000000
max       97.000000
Name: acousticness_%, dtype: float64

#### The average acousticness is 26.4. Songs in the Dataframe tend to be less acoustic.

### instrumentalness_%

In [42]:
songs_final_df['instrumentalness_%'].describe()

count    500.000000
mean       1.620000
std        8.888059
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max       91.000000
Name: instrumentalness_%, dtype: float64

#### The average instrumentalness is 1.6. Songs in the Dataframe have a great tendency to have a lead voice and not be only instrumental.

### liveness_%

In [43]:
songs_final_df['liveness_%'].describe()

count    500.000000
mean      17.462000
std       12.672084
min        3.000000
25%        9.000000
50%       12.000000
75%       22.000000
max       90.000000
Name: liveness_%, dtype: float64

#### The average liveness is 17.4. Songs in the DataFrame aren't usually recorded live.

### speechiness_%

In [44]:
songs_final_df['speechiness_%'].describe()

count    500.000000
mean       8.976000
std        8.350502
min        2.000000
25%        4.000000
50%        5.000000
75%       10.000000
max       45.000000
Name: speechiness_%, dtype: float64

#### The average speechiness is 8.9. Songs in the DataFrame don't tend to have spoken voices.

## Duration

In [45]:
songs_final_df['duration'].describe()

count    495.000000
mean     209.187727
std       54.314644
min       44.642000
25%      176.336000
50%      202.395000
75%      232.875000
max      613.026000
Name: duration, dtype: float64

#### The average duration of the songs is 209.18 seconds.

## Artist name

In [46]:
songs_final_df['artist_1_name'].value_counts()

artist_1_name
Bad Bunny                       23
The Weeknd                      19
Taylor Swift                    19
Harry Styles                    11
Eminem                           8
SZA                              8
Feid                             7
Doja Cat                         7
Drake                            7
Bruno Mars                       6
Olivia Rodrigo                   6
Ed Sheeran                       6
BTS                              6
Imagine Dragons                  5
Labrinth                         5
Justin Bieber                    5
Billie Eilish                    5
Dua Lipa                         5
Arctic Monkeys                   4
Kendrick Lamar                   4
Karol G                          4
Coldplay                         4
Post Malone                      4
Adele                            4
Ariana Grande                    4
Rauw Alejandro                   3
Lil Nas X                        3
Ti��sto                      3
J Balv

## genre

In [47]:
songs_final_df['genre'].value_counts()

genre
None                                                                                                     170
pop                                                                                                       46
reggaeton, trap latino, urbano latino                                                                     19
canadian contemporary r&b, canadian pop, pop                                                              11
dance pop, pop                                                                                            11
pop, r&b, rap                                                                                              7
                                                                                                           6
art pop, pop                                                                                               6
pop, singer-songwriter pop, uk pop                                                                         6
k-pop, k-pop 

#### The most repeated genre is pop, followed by reggaeton

# Relationships between variables

## Energy and danceability per released year

In [48]:
yearly_energy_danceability = songs_final_df.groupby('released_year')[['energy_%', 'danceability_%']].mean()
yearly_energy_danceability

energy_%  danceability_%
released_year                           
1942           25.000000       23.000000
1946           15.000000       36.000000
1950           32.000000       60.000000
1952           36.000000       67.000000
1957           37.000000       74.000000
1958           36.500000       71.500000
1959           24.000000       45.000000
1963           68.500000       38.500000
1968           70.000000       74.000000
1970           82.000000       50.000000
1971           61.000000       33.000000
1973           43.000000       39.000000
1975           56.500000       44.500000
1979           58.000000       75.000000
1983           45.000000       82.000000
1984           74.750000       65.750000
1985           68.000000       63.500000
1986           70.500000       52.500000
1987           90.000000       45.000000
1991           55.500000       48.000000
1992           34.000000       53.000000
1994           63.000000       34.000000
1995           72.500000       70.000000
1998           72.000000       32.000000
1999           76.200000       77.400000
2000           61.000000       62.333333
2002           69.166667       73.833333
2003           89.500000       42.500000
2004           69.000000       68.500000
2007           85.000000       52.000000
2008           63.500000       64.000000
2010           73.285714       61.714286
2011           63.100000       60.700000
2012           60.666667       61.555556
2013           62.727273       54.909091
2014           65.461538       63.538462
2015           61.400000       50.900000
2016           65.062500       60.562500
2017           59.380952       65.000000
2018           50.900000       55.000000
2019           63.133333       60.833333
2020           66.771429       68.371429
2021           63.362637       68.549451
2022           63.876543       70.419753
2023           65.277778       68.611111

### Correlation between energy and streams

In [49]:
# Calcular la correlación entre 'energy' y 'streams'
energy_streams_correlation = songs_final_df['energy_%'].corr(songs_final_df['streams'])

# Mostrar la correlación
print("Correlación entre energía y streams:", energy_streams_correlation)

Correlación entre energía y streams: 0.0055981689849439265


### Evolution on collaborations between artists

In [52]:
collaborations = songs_final_df.groupby('released_year')['artist_count'].value_counts()
collaborations

released_year  artist_count
1942           3                1
1946           1                1
1950           2                1
1952           1                1
1957           1                1
1958           1                1
               3                1
1959           1                1
1963           1                2
1968           1                1
1970           1                1
1971           4                1
1973           1                1
1975           1                2
1979           1                1
1983           1                1
1984           1                4
1985           1                2
1986           1                2
1987           1                1
1991           1                2
1992           1                1
1994           1                1
1995           2                1
               3                1
1998           1                1
1999           2                4
               1                1
2000           1    

In [53]:
colaborations_unstacked = collaborations.unstack().fillna(0)
colaborations_unstacked

artist_count      1     2     3    4    5    7
released_year                                 
1942            0.0   0.0   1.0  0.0  0.0  0.0
1946            1.0   0.0   0.0  0.0  0.0  0.0
1950            0.0   1.0   0.0  0.0  0.0  0.0
1952            1.0   0.0   0.0  0.0  0.0  0.0
1957            1.0   0.0   0.0  0.0  0.0  0.0
1958            1.0   0.0   1.0  0.0  0.0  0.0
1959            1.0   0.0   0.0  0.0  0.0  0.0
1963            2.0   0.0   0.0  0.0  0.0  0.0
1968            1.0   0.0   0.0  0.0  0.0  0.0
1970            1.0   0.0   0.0  0.0  0.0  0.0
1971            0.0   0.0   0.0  1.0  0.0  0.0
1973            1.0   0.0   0.0  0.0  0.0  0.0
1975            2.0   0.0   0.0  0.0  0.0  0.0
1979            1.0   0.0   0.0  0.0  0.0  0.0
1983            1.0   0.0   0.0  0.0  0.0  0.0
1984            4.0   0.0   0.0  0.0  0.0  0.0
1985            2.0   0.0   0.0  0.0  0.0  0.0
1986            2.0   0.0   0.0  0.0  0.0  0.0
1987            1.0   0.0   0.0  0.0  0.0  0.0
1991            2.0   0.0   0.0  0.0  0.0  0.0
1992            1.0   0.0   0.0  0.0  0.0  0.0
1994            1.0   0.0   0.0  0.0  0.0  0.0
1995            0.0   1.0   1.0  0.0  0.0  0.0
1998            1.0   0.0   0.0  0.0  0.0  0.0
1999            1.0   4.0   0.0  0.0  0.0  0.0
2000            3.0   0.0   0.0  0.0  0.0  0.0
2002            4.0   2.0   0.0  0.0  0.0  0.0
2003            2.0   0.0   0.0  0.0  0.0  0.0
2004            4.0   0.0   0.0  0.0  0.0  0.0
2007            1.0   0.0   0.0  0.0  0.0  0.0
2008            2.0   0.0   0.0  0.0  0.0  0.0
2010            6.0   1.0   0.0  0.0  0.0  0.0
2011            7.0   2.0   1.0  0.0  0.0  0.0
2012            7.0   2.0   0.0  0.0  0.0  0.0
2013           10.0   0.0   1.0  0.0  0.0  0.0
2014           13.0   0.0   0.0  0.0  0.0  0.0
2015           10.0   0.0   0.0  0.0  0.0  0.0
2016           11.0   4.0   1.0  0.0  0.0  0.0
2017           15.0   4.0   2.0  0.0  0.0  0.0
2018            7.0   3.0   0.0  0.0  0.0  0.0
2019           22.0   7.0   1.0  0.0  0.0  0.0
2020           26.0   7.0   2.0  0.0  0.0  0.0
2021           50.0  23.0  14.0  3.0  0.0  1.0
2022           99.0  47.0  14.0  1.0  1.0  0.0
2023            7.0  11.0   0.0  0.0  0.0  0.0